### PR TITLE
WIP - complete openSSL code

### PR DESCRIPTION
**Pull request by rvangraan**
TODO

- [ ] Import Certificate Chains
- [ ] Load Certificates from Keychain
- [ ] Test that Identity can be imported when Certificates are in the keychain

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,38 @@
+<<<<<<< HEAD
```
Comment (rymir) [inline .gitignore]: there are some conflicts here

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,38 @@
...
+# Directories potentially created on remote AFP share
+.AppleDB
+.AppleDesktop
+Network Trash Folder
+Temporary Items
+.apdisk
+
+
+### Objective-C ###
+>>>>>>> Import Existing Code
```
Comment (rymir) [inline .gitignore]: and here

File: IOSKeychain.podspec [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  #  These will help people to find your library, and whilst it
+  #  can feel like a chore to fill in it's definitely to your advantage. The
+  #  summary should be tweet-length, and the description more in depth.
+  #
+
+  s.name         = "IOSKeychain"
+  s.version      = "1.0.0"
+  s.summary      = "A Library encapsulating IOS keychain access and certificates"
+
+  s.description  = <<-DESC
```
Comment (rymir) [inline IOSKeychain.podspec]: add description here

File: IOSKeychain/Certificate.swift [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    public init(secCertificate: SecCertificate) {
+        super.init(securityClass: .Certificate, attributeBag: nil)
+        self.secCertificate = secCertificate
+        let unManagedSubjectSummary = SecCertificateCopySubjectSummary(secCertificate)
+        self.subject = unManagedSubjectSummary.takeRetainedValue() as String
+    }
+}
+
+//public class Certificate: KeyChainItem {
```
Comment (rymir) [inline IOSKeychain/Certificate.swift]: remove commented code

File: IOSKeychain/IOSKeychain.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+//! Project version number for IOSKeychain.
+FOUNDATION_EXPORT double IOSKeychainVersionNumber;
+
+//! Project version string for IOSKeychain.
+FOUNDATION_EXPORT const unsigned char IOSKeychainVersionString[];
+
+// In this header, you should import all the public headers of your framework using statements like #import <IOSKeychain/PublicHeader.h>
+
+
+#import <IOSKeychain/OpenSSL.h>
```
Comment (rymir) [inline IOSKeychain/IOSKeychain.h]: missing new line at the end of the file

File: IOSKeychain/Identity.swift [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public var p12EncodedIdentityData: NSData {
+        get { return openSSLIdentity.p12identityData }
+    }
+}
+
+public class Identity : KeyPair {
+
+}
```
Comment (rymir) [inline IOSKeychain/Identity.swift]: missing newline

File: IOSKeychain/ItemSpecifier.swift [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+public protocol KeyPairQuery : KeychainQuery {
+    func privateKeySpecifier() -> KeySpecifier
+    func publicKeySpecifier() -> KeySpecifier
+}
+
+
+
+public class KeySpecification : KeySpecifier {
+    public class func specification(PEMEncodedPrivateKey privateKeyData: NSData, passphrase: String, keyLabel: String , keyAppTag: String? = nil, keyAppLabel: String? = nil) -> KeyPairImportSpecification? {
```
Comment (rymir) [inline IOSKeychain/ItemSpecifier.swift]: this line is more then 173 chars long, can we make it 100?

File: IOSKeychain/Keychain.swift [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+        if status == .OK {
+            println(persistedRef)
+            if let data = persistedRef as? NSData {
+                println(data)
+            }
+        }
+
+
+        return (.OK,[])
+//        println(status)
```
Comment (rymir) [inline IOSKeychain/Keychain.swift]: remove commented code

File: IOSKeychain/Keychain.swift [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+
+        let status = KeychainStatus.statusFromOSStatus(
+            withUnsafeMutablePointer(&persistedRef) { SecItemAdd(attributes, UnsafeMutablePointer($0)) }
+        )
+
+        if status == .OK {
+            println(persistedRef)
+            if let data = persistedRef as? NSData {
+                println(data)
```
Comment (rymir) [inline IOSKeychain/Keychain.swift]: remove println

File: IOSKeychain/Keychain.swift [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+            let identityRef : SecIdentityRef = itemRefs[0][kSecImportItemIdentity.takeRetainedValue() as String] as! SecIdentityRef
+            return IdentityReference(reference: identityRef)
+        }
+        return nil
+    }
+
+    public class func deleteAllItemsOfClass(securityClass: SecurityClass) -> Int {
+        return 0
+    }
+}
```
Comment (rymir) [inline IOSKeychain/Keychain.swift]: missing new line, can you set the editor to add new lines at the end of the files?

File: IOSKeychain/OpenSSL.m [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+//
+//  Created by Rudolph van Graan on 23/06/2014.
+//  Copyright (c) 2014 Curoo Limited. All rights reserved.
+//
+
+
+#import <IOSKeychain/IOSKeychain-Swift.h>
+
+#import "OpenSSL.h"
+//#import "ExKCCertificate.h"
```
Comment (rymir) [inline IOSKeychain/OpenSSL.m]: remove commented code

File: IOSKeychainTests/CertificateTests.swift [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+class CertificateTests: XCTestCase {
+
+    func testCertificateFromCERFile() {
+
+        let bundle = NSBundle(forClass: self.dynamicType)
+
+        let certificateDERData : NSData! = NSData(contentsOfFile: bundle.pathForResource("Curoo Root CA", ofType: "cer")!)
+
+        let certificate : Certificate! = Certificate.certificate(certificateDERData)
```
Comment (rymir) [inline IOSKeychainTests/CertificateTests.swift]: why is this a class fun and not `let certificate: Certificate! = Certificate(certificateDERData)`?

File: IOSKeychainTests/KeyPairTests.swift [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        XCTAssertEqual(count(items),0)
+
+        let keyPairSpecifier = PermanentKeyPairSpecification(keyType: .RSA, keySize: 1024, keyLabel: "AAA", keyAppTag: "BBB", keyAppLabel: "CCC")
+        var keyPair : KeyPair?
+        (status, keyPair) = Keychain.generateKeyPair(keyPairSpecifier)
+        XCTAssertEqual(status, .OK)
+        XCTAssertNotNil(keyPair)
+
+        (status, items) = Keychain.keyChainItems(.Key)
+        XCTAssertEqual(count(items),2)
```
Comment (rymir) [inline IOSKeychainTests/KeyPairTests.swift]: so a keypair generates 2 items in the keychain? pub and private keys?

File: IOSKeychainTests/KeyPairTests.swift [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+        let keySpecifier = KeySpecifier(keyLabel: "AAA")
+        var keyItem: KeychainItem?
+        (status, keyItem) = Keychain.fetchMatchingItem(itemSpecifier: keySpecifier)
```
Comment (rymir) [inline IOSKeychainTests/KeyPairTests.swift]: cause here you treat it as a single item, so I was expecting a count of 1 in the comment before

File: IOSKeychainTests/KeyPairTests.swift [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        XCTAssertEqual(publicKeyData!.length,140)
+        XCTAssert(privateKeyData!.length > 0)
+
+    }
+
+    func testGenerateUnnamedKeyPair() {
+        clearKeychainItems(.Key)
+        var (status, items) = Keychain.keyChainItems(.Key)
+        XCTAssertEqual(status, .OK)
+        XCTAssertEqual(count(items),0)
```
Comment (rymir) [inline IOSKeychainTests/KeyPairTests.swift]: these last 4 lines are always the same, group them into a function

**Comment by rvangraan** [comment]
Obsolete.

**Pull request closed without merging**